# Regression Models with Keras



## Introduction


Despite the popularity of more powerful libraries such as PyToch and TensorFlow, they are not easy to use and have a steep learning curve. So, for people who are just starting to learn deep learning, there is no better library to use other than the Keras library. 

Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development. As you will see in this assignment, we will build a regression model using the deep learning Keras library, and then we will experiment with increasing the number of training epochs and changing number of hidden layers and see how changing these parameters impacts the performance of the model. 


<h2>Regression Models with Keras</h2>

<h3>Objective for this Notebook<h3>    
<h5> 1. In this project, you will build a regression model using the Keras library to model the same data about concrete compressive strength that we used in labs 3.l.</h5>
<h5> 2. Download and Clean dataset:  
https://cocl.us/concrete_data </h5>
<h5> 3. Build a Neural Network </h5>
<h5> 4. Train and Test the Network. </h5>     



## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
1. <a href="#item31">Download and Clean concrete compressive strength Dataset</a>  
2. <a href="#item32">Import Keras</a>  
3. <a href="#item33">Build a Neural Network and then experiment with increasing the number of training epochs and changing number of hidden layers</a>  
4. <a href="#item34">Train and Test the Network, note impact on performance on MSE with change in hyperparameter</a>  

</font>
</div>


<a id="item31"></a>


## Download and Clean Dataset


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [18]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented. 
# If you run this notebook on a different environment, e.g. your desktop, you may need to uncomment and install certain libraries.

!pip install tensorflow
#!pip install pandas==1.3.4
!pip install keras

Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, we will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. 

In [19]:
import pandas as pd
import numpy as np
import tensorflow as tf
import keras

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [20]:
from keras.models import Sequential
from keras.layers import Dense

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [21]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 


#### Let's check how many data points we have.


In [22]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [23]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [24]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [25]:
concrete_data_columns = concrete_data.columns

X = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
y = concrete_data['Strength'] # Strength column

<a id="item2"></a>


Let's do a quick sanity check of the predictors and the target dataframes.


In [26]:
X.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [27]:
y.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.


In [28]:
X_norm = (X - X.mean()) / X.std()
X_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to *n_cols* since we will need this number when building our network.


In [29]:
n_cols = X_norm.shape[1] # number of predictors

<a id="item1"></a>


<a id='item32'></a>


## Build a Neural Network


Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.


In [30]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has two hidden layers, each of 50 hidden units.


<a id="item4"></a>


<a id='item34'></a>


## Train and Test the Network


Let's call the function now to create our model.


In [31]:
# build the model
model = regression_model()

Next, we will train and test the model at the same time using the *fit* method. We will leave out 30% of the data for validation and we will train the model for 100 epochs.


In [32]:
# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)

Epoch 1/100
23/23 - 2s - loss: 1682.2167 - val_loss: 1214.3463 - 2s/epoch - 71ms/step
Epoch 2/100
23/23 - 0s - loss: 1655.9927 - val_loss: 1196.9022 - 154ms/epoch - 7ms/step
Epoch 3/100
23/23 - 0s - loss: 1625.5348 - val_loss: 1176.0179 - 141ms/epoch - 6ms/step
Epoch 4/100
23/23 - 0s - loss: 1587.6316 - val_loss: 1150.6724 - 162ms/epoch - 7ms/step
Epoch 5/100
23/23 - 0s - loss: 1540.5710 - val_loss: 1119.8024 - 167ms/epoch - 7ms/step
Epoch 6/100
23/23 - 0s - loss: 1480.4769 - val_loss: 1082.0232 - 152ms/epoch - 7ms/step
Epoch 7/100
23/23 - 0s - loss: 1405.2887 - val_loss: 1035.3990 - 151ms/epoch - 7ms/step
Epoch 8/100
23/23 - 0s - loss: 1313.5629 - val_loss: 978.7112 - 142ms/epoch - 6ms/step
Epoch 9/100
23/23 - 0s - loss: 1202.8196 - val_loss: 910.6013 - 129ms/epoch - 6ms/step
Epoch 10/100
23/23 - 0s - loss: 1075.4708 - val_loss: 833.4981 - 133ms/epoch - 6ms/step
Epoch 11/100
23/23 - 0s - loss: 937.5427 - val_loss: 752.4839 - 133ms/epoch - 6ms/step
Epoch 12/100
23/23 - 0s - loss: 799.8

In [33]:
# Evaluate the model on the test data
evaluation = model.evaluate(predictors_norm, target)

# Print the evaluation results
print("Loss:", evaluation)  # The loss value (e.g., MSE) will be printed

33/33 [==============================] - 0s 2ms/step - loss: 129.4616
Loss: 129.46157836914062


In [34]:
# Calculate the MSE manually
y_pred = model.predict(X_norm)

# Ensure that both y_pred and target are 1-dimensional arrays
# The flatten() method is not present on a Pandas Series object,to convert a Pandas Series to a 1-dimensional NumPy array, use the .values attribute . 
# y_pred = y_pred.flatten()
# target = target.flatten()

# Convert Pandas Series to 1-dimensional NumPy arrays
y_pred = y_pred.flatten()
y = y.values

mse = np.mean((y_pred - y) ** 2)

print( " Mean Sqaured Error: ", mse)


33/33 [==============================] - 0s 2ms/step
 Mean Sqaured Error:  129.46158379444256
